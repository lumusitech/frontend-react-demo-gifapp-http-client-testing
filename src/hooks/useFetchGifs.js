import { useState, useEffect } from 'react'

import { getGifs } from '../helpers/getGifs'

export const useFetchGifs = (category) => {
    const [state, setState] = useState({
        data: [],
        loading: true
    })

    // useEffect(() => {
    //     getGifs(category)
    //         .then(setImages); // Forma abreviada
    //     // Cuando solo se pasa un elemento, se simplifica
    //     // poniendo solamente el nombre de la función
    //     // Forma extensa
    //     // .then(imgs => setImages(imgs));
    // }, [category]); // Si está en la lista de cambios, se renderizará

    useEffect(() => {
        getGifs(category)
            .then(imgs => {
                setState({
                    data: imgs,
                    loading: false
                })
            })
    }, [category])


    return state; // { data: [], loading: true }
}