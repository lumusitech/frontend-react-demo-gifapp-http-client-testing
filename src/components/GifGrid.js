import React from 'react'
import { useFetchGifs } from '../hooks/useFetchGifs'
import PropTypes from 'prop-types'

import { GifGridItem } from './GifGridItem'

const GifGrid = ({ category }) => {

    const { data:images, loading } = useFetchGifs(category)

    return (
        <>
            <h3>{category}</h3>

            {loading && <p className="card-grid animate__animated animate__flash">Loading...</p>}

            <div className="card-grid animate__animated animate__fadeIn animate__delay-1s">
                {
                    images.map(img => (
                        <GifGridItem
                            key={img.id}
                            {...img}
                        />
                    )
                    )
                }
            </div>
        </>
    )
}

export default GifGrid

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}