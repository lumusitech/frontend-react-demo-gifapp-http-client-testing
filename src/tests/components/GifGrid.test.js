import React from 'react'
import { shallow } from 'enzyme'
import '@testing-library/jest-dom'
import GifGrid from '../../components/GifGrid'
import { useFetchGifs } from '../../hooks/useFetchGifs'
jest.mock('../../hooks/useFetchGifs')

describe('Testing on GifGrid.js', () => {

    const category = 'Hola'

    test('should match with snapshot along time', () => {
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        })
        const wrapper = shallow(<GifGrid category={category} />);
        expect(wrapper).toMatchSnapshot();
    })

    test('should shows items when images already loading', () => {
        const gifs = [
            {
                id: 1,
                title: 'Cualquier cosa',
                url: 'https://localhost/cualquier/cosa.jpg'
            },
            {
                id: 2,
                title: 'otra cosa',
                url: 'https://localhost/otra/cosa.jpg'
            }
        ];
        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        })

        const wrapper = shallow(<GifGrid category={category} />);

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('p').exists()).toBeFalsy();
        expect(wrapper.find('GifGridItem').length).toBe(gifs.length);
    })
})