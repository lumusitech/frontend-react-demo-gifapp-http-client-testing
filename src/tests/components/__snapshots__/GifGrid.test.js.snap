// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing on GifGrid.js should match with snapshot along time 1`] = `
<Fragment>
  <h3>
    Hola
  </h3>
  <p
    className="card-grid animate__animated animate__flash"
  >
    Loading...
  </p>
  <div
    className="card-grid animate__animated animate__fadeIn animate__delay-1s"
  />
</Fragment>
`;

exports[`Testing on GifGrid.js should shows items when images already loading 1`] = `
<Fragment>
  <h3>
    Hola
  </h3>
  <div
    className="card-grid animate__animated animate__fadeIn animate__delay-1s"
  >
    <GifGridItem
      id={1}
      key="1"
      title="Cualquier cosa"
      url="https://localhost/cualquier/cosa.jpg"
    />
    <GifGridItem
      id={2}
      key="2"
      title="otra cosa"
      url="https://localhost/otra/cosa.jpg"
    />
  </div>
</Fragment>
`;
